File: 2020-week15/sw.cpp
/*
1. queries 탐색
2. query 1개를 탐색할 때마다 words를 돌면서 탐색
    - word가 해당 query와 매치되는지 검사
    - 매치되면 갯수 카운트
3. 카운트를 result에 넣어줌

매치 검사 logic
0. 문자 길이 먼저 검사
1. 문자열 탐색
2. 알파벳이 같을 경우 통과
3. 알파벳이 다를 경우 탐색 종료
4. 와일드카드 문자는 무조건 통과
*/

#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> words, vector<string> queries) {
    vector<int> answer;
    
    for(string query: queries){
        int count = 0;
        
        for(string word: words){
            bool flag = true;
            if(query.size() != word.size()) continue;
            
            for(int i=0; i<query.size(); i++){
                if(query[i] == '?')     continue;
                
                if(query[i] != word[i]){
                    flag = false;
                    break;
                }
            }
            
            if(flag)    count++;
        }
        
        answer.push_back(count);
    }
    return answer;
}

File: 2020-week15/sw.js
class Node {
  constructor() {
    this.count = 0;
    this.child = {};
  }
}

class Trie {
  constructor(word) {
    this.root = new Node();
  }

  // Trie에 삽입
  insert(string) {
    let currentNode = this.root;

    for (let i = 0; i < string.length; i++) {
      const currentChar = string[i];

      if (currentNode.child[currentChar] === undefined) {
        currentNode.child[currentChar] = new Node();
      }

      currentNode.count += 1;
      currentNode = currentNode.child[currentChar];
    }
  }

  // 검색
  search(string) {
    let currentNode = this.root;

    for (let i = 0; i < string.length; i++) {
      const currentChar = string[i];

      if (currentChar === "?") {
        break;
      }

      if (currentNode.child[currentChar] !== undefined) {
        currentNode = currentNode.child[currentChar];
      } else {
        return 0;
      }
    }
    return currentNode.count;
  }
}

function solution(words, queries) {
  const answer = [];

  const trie = {};
  const reverseTrie = {};

  // trie 생성
  words.forEach((word) => {
    if (trie[word.length] === undefined) {
      trie[word.length] = new Trie();
      reverseTrie[word.length] = new Trie();
    }

    trie[word.length].insert(word);
    reverseTrie[word.length].insert(word.split("").reverse().join(""));
  });

  // queries 탐색
  queries.forEach((query) => {
    // 접두사일 때 => trie 탐색
    if (query[0] === "?") {
      if (reverseTrie[query.length] === undefined) {
        answer.push(0);
      } else {
        answer.push(
          reverseTrie[query.length].search(query.split("").reverse().join(""))
        );
      }
    }
    // 접미사일 때 => reverseTrie 탐색
    else {
      if (trie[query.length] === undefined) {
        answer.push(0);
      } else {
        answer.push(trie[query.length].search(query));
      }
    }
  });

  return answer;
}

const words = ["frodo", "front", "frost", "frozen", "frame", "kakao"];
const queries = ["fro??", "????o", "fr???", "fro???", "pro?"];
console.log(solution(words, queries));
